### PR TITLE
nixos/backups/restic: handle cases when both dynamicFileFrom and paths are set

### PR DESCRIPTION
**Pull request by robryk**
Also, add a test to verify that it works.

This change also removes the part of custom package test that verifies that the correct paths are provided. This is already tested by restore tests.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@i077

### CODE REVIEW

**Comment by robryk** [comment]
Gentle ping @bbigras @i077

**Comment by robryk** [comment]
Ping? The delay is forcing me to keep rebasing this.

**Review by i077 — CHANGES_REQUESTED** [review]
Apologies for the delay! This looks fine to me, just think the options' docs should clarify the module's behavior.

File: nixos/modules/services/backup/restic.nix [hunk 1]
```diff
@@ -113,8 +113,10 @@ in
...
 
         paths = mkOption {
+          # This is nullable for legacy reasons only. We should consider making it a pure listOf
+          # after some time has passed since this comment was added.
           type = types.nullOr (types.listOf types.str);
-          default = null;
+          default = [ ];
           description = lib.mdDoc ''
             Which paths to backup.  If null or an empty array, no
             backup command will be run.  This can be used to create a
```
[the remark this answers is not in the document]
Comment (robryk) [reply nixos/modules/services/backup/restic.nix]: Done.

**Comment by robryk** [comment]
Ping?

**Review by i077 — APPROVED** [review]
LGTM!

**Comment by pbsds** [comment]
@ofborg test restic

**Comment by Janik-Haag** [comment]
I think this broke restic on unstable. I get this error: 
image
since updating from 63678e9f3d3afecfeafa0acead6239cdb447574c to 0cbe9f69c234a7700596e943bfae7ef27a31b735
and the only other update to the module was #263748 which seems unrelated and my restic config didn't change

**Comment by robryk** [comment]
I got an email notification about a comment here from "Janik" (with a breakage report), but I don't see the comment or any trace thereof here. Does anyone know what's going on?

**Comment by hughobrien** [comment]
This is inserting `--files-from=/run/<job>/includes` which doesn't exist in my config (I specify a path to scan). Jobs are broken as a result.

**Comment by MinerSebas** [comment]
> This is inserting `--files-from=/run/<job>/includes` which doesn't exist in my config (I specify a path to scan). Jobs are broken as a result.
> 
> I think
> 
> ```
>  ExecStart = (optionals doBackup [ "${resticCmd} backup ${concatStringsSep " " (backup.extraBackupArgs ++ excludeFlags)} --files-from=${filesFromTmpFile}" ])
> ```
> 
> needs to only include `files-from` when it is defined.

I think the actual correct fix is adding a `|| (backup.paths != null && backup.paths != [])` to this Block:
https://github.com/NixOS/nixpkgs/blob/85f1ba3e51676fa8cc604a3d863d729026a6b8eb/nixos/modules/services/backup/restic.nix#L348-L362
so that the static paths are actually set.

**Comment by Janik-Haag** [comment]
@MinerSebas please take a look at  #264696 and the linked PRs

**Pull request merged**